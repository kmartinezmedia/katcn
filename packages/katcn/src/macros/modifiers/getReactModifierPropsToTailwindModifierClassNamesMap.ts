import { SyntaxKind } from 'ts-morph';

import { getTwPrefixAndPropName } from '../getTwPrefixAndPropName';
import { katcnTypesSourceFile } from '../tsMorph';

/**
 * Used in tooling to convert react modifier props to tailwind classnames
 *
 * This lookup map be used in conjuction with other ts-morph tooling
 * to convert to <Box _hover={{ bg: 'red-500' }} /> to
 * `<Box className="hover:bg-red-500" />`
 *
 * Returns:
 *  { _hover: 'hover', ...other modifiers }
 */
export function getReactModifierPropsToTailwindModifierClassNamesMap() {
  const lookupMap = {} as Record<string, string>;
  const modifierPropsTypeDefinition =
    katcnTypesSourceFile
      .getInterface('ModifierProps')
      ?.getChildrenOfKind(SyntaxKind.PropertySignature) ?? [];

  for (const propertySignature of modifierPropsTypeDefinition) {
    if (!propertySignature) continue;
    const { name, twPrefix } = getTwPrefixAndPropName(propertySignature);
    lookupMap[name] = twPrefix;
  }
  return lookupMap;
}
