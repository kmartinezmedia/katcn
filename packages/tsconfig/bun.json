{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Bun",
  "extends": "./node.json",
  "compilerOptions": {
    "types": ["bun-types"]
  }
}
