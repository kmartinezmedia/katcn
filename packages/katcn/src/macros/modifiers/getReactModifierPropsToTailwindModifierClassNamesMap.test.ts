import { describe, expect, it } from 'bun:test';
import { getReactModifierPropsToTailwindModifierClassNamesMap } from './getReactModifierPropsToTailwindModifierClassNamesMap';

describe('getReactModifierPropsToTailwindModifierClassNamesMap', () => {
  it('Matches snapshot', () => {
    const value = getReactModifierPropsToTailwindModifierClassNamesMap();
    expect(value).toMatchSnapshot();
  });
  it('Should return a mapping of katcn modifier props to tailwind modifier classnames', () => {
    const value = getReactModifierPropsToTailwindModifierClassNamesMap();
    expect(value).toBeDefined();
    expect(value).toMatchObject({
      _hover: 'hover:',
    });
  });
});
