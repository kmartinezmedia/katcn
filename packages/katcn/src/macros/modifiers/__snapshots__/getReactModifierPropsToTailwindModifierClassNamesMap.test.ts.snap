// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getReactModifierPropsToTailwindModifierClassNamesMap Matches snapshot 1`] = `
{
  "_2xl": "2xl:",
  "_active": "active:",
  "_after": "after:",
  "_ariaBusy": "aria-busy:",
  "_ariaChecked": "aria-checked:",
  "_ariaDisabled": "aria-disabled:",
  "_ariaExpanded": "aria-expanded:",
  "_ariaHidden": "aria-hidden:",
  "_ariaPressed": "aria-pressed:",
  "_ariaReadonly": "aria-readonly:",
  "_ariaRequired": "aria-required:",
  "_ariaSelected": "aria-selected:",
  "_autofill": "autofill:",
  "_backdrop": "backdrop:",
  "_checked": "checked:",
  "_cmdkGroup": "cmdkGroup:",
  "_cmdkGroupHeading": "cmdkGroupHeading:",
  "_cmdkGroupItems": "cmdkGroupItems:",
  "_cmdkInput": "cmdkInput:",
  "_cmdkInputSvg": "cmdkInputSvg:",
  "_cmdkItem": "cmdkItem:",
  "_cmdkItemSvg": "cmdkItemSvg:",
  "_contrastLess": "contrast-less:",
  "_contrastMore": "contrast-more:",
  "_dark": "dark:",
  "_dataDisabled": "data-[disabled=true]:",
  "_default": "default:",
  "_disabled": "disabled:",
  "_empty": "empty:",
  "_enabled": "enabled:",
  "_even": "even:",
  "_file": "file:",
  "_first": "first:",
  "_firstOfType": "first-of-type:",
  "_focus": "focus:",
  "_focusVisible": "focus-visible:",
  "_focusWithin": "focus-within:",
  "_forcedColors": "forced-colors:",
  "_hover": "hover:",
  "_inRange": "in-range:",
  "_indeterminate": "indeterminate:",
  "_invalid": "invalid:",
  "_last": "last:",
  "_lastOfType": "last-of-type:",
  "_lg": "lg:",
  "_md": "md:",
  "_motionReduce": "motion-reduce:",
  "_motionSafe": "motion-safe:",
  "_odd": "odd:",
  "_only": "only:",
  "_onlyOfType": "only-of-type:",
  "_outOfRange": "out-of-range:",
  "_placeholder": "placeholder:",
  "_placeholderShown": "placeholder-shown:",
  "_readonly": "read-only:",
  "_required": "required:",
  "_selection": "selection:",
  "_sm": "sm:",
  "_target": "target:",
  "_valid": "valid:",
  "_visited": "visited:",
  "_xl": "xl:",
}
`;
